### PR TITLE
KAFKA-18001: Support UpdateRaftVoterRequest in KafkaNetworkChannel

### PR DESCRIPTION
**Pull request by justinrlee**
Small fix: Update KafkaNetworkChannel to support UpdateRaftVoterRequest.

Addresses this scenario:
* Bootstrap a KRaft Controller quorum in dynamic mode
* Start additional controllers (as observers)
* Update `kraft.version` feature from 0 to 1
* Use `kafka-metadata-quorum add-controller` to promote an observer controller to a follower

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ahuang98** [comment]
cc @jsancio

**Comment by ahuang98** [comment]
Thanks @justinrlee, created https://issues.apache.org/jira/browse/KAFKA-18001 to track, can you prepend your PR title with `KAFKA-18001:`?

**Review by ahuang98 — CHANGES_REQUESTED** [review]
lgtm, as for testing, `KafkaNetworkChannelTest` could be the place for it

**Comment by justinrlee** [comment]
Thanks, @ahuang98! I have updated the PR title and will look at adding tests today or tomorrow.

**Comment by justinrlee** [comment]
Okay, @ahuang98 , I've added tests similar to the existing tests in KafkaNetworkChannelTest; please let me know what else you need from me!

File: raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java
Comment (OmniaGM) [inline raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java:192]: I believe `RemoveRaftVoterRequestData` needs to be added as well.

**Review by ahuang98 — APPROVED** [review]
thanks for the additional test file changes!

File: raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java
Comment (cmccabe) [reply raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java:192]: Do the controllers send `RemoveRaftVoterRequestData` and `AddRaftVoterRequestData` here? I don't think so?

**Review by cmccabe — APPROVED** [review]
LGTM

File: raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java:192]: that's what I thought as well, I don't think we need to add either here

**Comment by justinrlee** [comment]
Looks like all the tests are failing due to my new tests bumping us over the threshold (25)for the `ClassDataAbstractionCoupling` checkstyle:

```
Error: eckstyle] [ERROR] /home/runner/work/kafka/kafka/raft/src/test/java/org/apache/kafka/raft/KafkaNetworkChannelTest.java:68:1: Class Data Abstraction Coupling is 26 (max allowed is 25) classes [AssertionError, BeginQuorumEpochResponse, BeginQuorumEpochResponseData, EndQuorumEpochResponse, EndQuorumEpochResponseData, FetchRequestData.ReplicaState, FetchResponse, FetchResponseData, FetchSnapshotResponse, FetchSnapshotResponseData, KafkaNetworkChannel, MockClient, MockTime, Node, OffsetAndEpoch, RaftRequest.Outbound, StubMetadataUpdater, SupportedVersionRange, Thread, TopicPartition, UpdateRaftVoterResponse, UpdateRaftVoterResponseData, VoteResponse, VoteResponseData, VoteResponseData.PartitionData, VoteResponseData.TopicData]. [ClassDataAbstractionCoupling]
```

Two questions:
* Is it acceptable to just suppress this warning? I assume so, because I see a ton of suppressions for similar test cases across the project.
* Is it preferable to suppress in the suppress in the global [checkstyle suppressions file](https://github.com/apache/kafka/blob/trunk/checkstyle/suppressions.xml), or is it preferable to do an inline suppression (I see a [handful](https://github.com/apache/kafka/blob/4fcb8b247ec28b9f6e64f8f663e34f5becf0f25f/tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java#L77) of [these](https://github.com/apache/kafka/blob/4fcb8b247ec28b9f6e64f8f663e34f5becf0f25f/connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java#L86) [across](https://github.com/apache/kafka/blob/4fcb8b247ec28b9f6e64f8f663e34f5becf0f25f/group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupCoordinatorRecordSerdeTest.java#L67) [the](https://github.com/apache/kafka/blob/4fcb8b247ec28b9f6e64f8f663e34f5becf0f25f/storage/src/test/java/org/apache/kafka/tiered/storage/TieredStorageTestBuilder.java#L70) [project](https://github.com/apache/kafka/blob/4fcb8b247ec28b9f6e64f8f663e34f5becf0f25f/clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessorTest.java#L77))?

File: raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java
Comment (justinrlee) [reply raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java:192]: FWIW, I don't see any errors when I use `remove-controller` so it doesn't look like we need to support this.

**Comment by justinrlee** [comment]
@ahuang98 

Added checkstyle suppression to the global suppression list; let me know if you need me to do something else!

Also, I think theoretically we want this both in `trunk` and `3.9` but from reading the [contribution guide](https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Code+Changes) I don't quite understand whether I need to do that manually or if there's some other process for this.

**Comment by ahuang98** [comment]
@justinrlee thanks for the fix! getting this into 3.9 would be a manual step - after this is merged into trunk, you can open up a PR targeting the `3.9` branch with your cherry-pick (and tag @cmccabe to merge it in)

**Comment by OmniaGM** [comment]
LGTM, it would be nice if we can add some ducktape test for the KIP at some point. I don't think this suggestion should block this PR

**Comment by ahuang98** [comment]
@OmniaGM we do have https://github.com/apache/kafka/blob/trunk/tests/kafkatest/tests/core/quorum_reconfiguration_test.py but you're right that it doesn't cover this case due to the quorum being brought up with kraft.version=1 already. The test Justin added is pretty straightforward in confirming UpdateRaftVoterRequests can make it through

**Pull request merged**
